tools/scripts/update-bumpenvs-yaml.py

```diff
@@ -0,0 +1,169 @@
+
+
+def get_all_artifacts(builds: Dict[str, Build]) -> Dict[str, str]:
+    artifacts = {}
+    for b in builds.values():
+        artifacts.update(b.get_artifacts())
+
+    found = set(artifacts.keys())
+    assert (
+        EXPECT_ARTIFACTS == found
+    ), f"expected artifacts ({EXPECT_ARTIFACTS}) but found ({found})"
+
+    return artifacts
+
+
+def parse_packer_log(packer_log: str) -> Dict[str, str]:
+    """Parse the output of packer's -machine-readable format, strange though it may be."""
+    out = {}
+
+    lines = packer_log.strip().split("\n")
+    fields = [line.split(",") for line in lines]
+    # We actually don't care about non-artifact lines or lines with fewer than 6 fields.
+    fields = [f for f in fields if len(f) > 5 and f[2] == "artifact"]
+
+    # Get the ami images, which should match lines like this one (line break for readability):
+    #   1598642161,amazon-ebs,artifact,0,id,us-east-1:
+    #       ami-04894a7352df9fdf9%!(PACKER_COMMA)us-west-2:ami-017627938fe327e4f
+    aws_fields = [f for f in fields if f[1] == "amazon-ebs"]
+    ami_lines = [f[5] for f in aws_fields if f[4] == "id" and f[5]]
+    assert (
+        len(ami_lines) == 1
+    ), f"expected one matching ami ids line but got: {ami_lines}"
+
+    ami_fields = ami_lines[0].split("%!(PACKER_COMMA)")
+    for ami_field in ami_fields:
+        region, ami = ami_field.split(":")
+        name = region.replace("-", "_") + "_agent_ami"
+        out[name] = ami
+
+    # Get the GCP builder name by matching the build-id from a line like this one:
+    #   1598642161,det-environments-06318c7,artifact,0,builder-id,packer.googlecompute
+    gcp_builders = [
+        f[1] for f in fields if f[4] == "builder-id" and f[5] == "packer.googlecompute"
+    ]
+    assert len(gcp_builders) == 1, f"expected one gcp builder but got: {gcp_builders}"
+    gcp_builder = gcp_builders[0]
+
+    # Get the GCP artifact ID by matching a line like this one:
+    #    1598642161,det-environments-06318c7,artifact,0,id,det-environments-06318c7
+    gcp_ids = [f[5] for f in fields if f[1] == gcp_builder and f[4] == "id" and f[5]]
+    assert len(gcp_ids) == 1, f"expected one matching gcp id line but got: {gcp_ids}"
+    out["gcp_env"] = gcp_ids[0]
+
+    return out
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 3:
+        print(__doc__, file=sys.stderr)
+        sys.exit(1)
+
+    path = sys.argv[1]
+    commit = sys.argv[2]
+
+    with open(path) as f:
+        conf = yaml.safe_load(f)
+
+    builds = get_all_builds(commit)
+    artifacts = get_all_artifacts(builds)
+
+    parse_packer_log(artifacts["packer-log"])
+
+    conf.update(yaml.safe_load(artifacts["publish-tf1-cpu"]))
+    conf.update(yaml.safe_load(artifacts["publish-tf1-gpu"]))
+    conf.update(yaml.safe_load(artifacts["publish-tf2-cpu"]))
+    conf.update(yaml.safe_load(artifacts["publish-tf2-gpu"]))
+    conf.update(parse_packer_log(artifacts["packer-log"]))
+
+    # Sort the yaml output to avoid git file churn.
```

Comment:
blocking: It looks like `yaml.dump` should be sorting keys already; have you found this logic to be necessary? You can pass `sort_keys=True` to make that totally explicit, which also avoids assuming that each field is one line.

Reply:
I guess you are right... it is already sorting.  I don't know where this `sort_keys=True` option is from though, I can't find it anywhere.